File: 백준/Silver/9012. 괄호/괄호.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int T = Integer.parseInt(br.readLine());
        for(int tc = 0; tc < T; tc++) {
            char[] arr = br.readLine().toCharArray();
            Stack<Character> stack = new Stack<>();
            boolean isTrue = true;
            for(int i = 0; i < arr.length; i++) {
                if(arr[i] == '(') {
                    stack.push('(');
                }
                else {
                    if(stack.isEmpty() || stack.peek() != '(') {
                        sb.append("NO\n");
                        isTrue = false;
                        break;
                    } else {
                        stack.pop();
                    }
                }
            }
            if(isTrue) sb.append(stack.isEmpty() ? "YES" : "NO").append("\n");
        }
        System.out.println(sb.toString().trim());
    }
}

File: 백준/Silver/9012. 괄호/괄호.js
class Stack {
  constructor() {
    this.arr = [];
  }

  push(num) {
    this.arr.push(num);
  }

  pop() {
    return this.arr.pop();
  }

  peek() {
    return this.arr[this.arr.length - 1];
  }

  size() {
    return this.arr.length;
  }

  isEmpty() {
    return this.arr.length === 0;
  }
}

let [T, ...input] = require('fs').readFileSync('/dev/stdin').toString().split('\n');
const answer = [];
input = input.map(line => line.trim().split(''));
T = Number(T);
for(let i = 0; i < T; i++) {
  let stack = new Stack();
  let isTrue = true;
  for(let j = 0; j < input[i].length; j++) {
    if(input[i][j] == '(') stack.push('(');
    else {
      if(stack.isEmpty() || stack.peek() !== '(') {
        isTrue = false;
        answer.push('NO');
        break;
      } else {
        stack.pop();
      }
    }
  }
  if(isTrue) answer.push(stack.isEmpty() ? 'YES' : 'NO');
}
console.log(answer.join('\n'));